### PR TITLE
[SPARK-19152][SQL]DataFrameWriter.saveAsTable support hive append

### PR DESCRIPTION
**Pull request by windpiger**
## What changes were proposed in this pull request?

After [SPARK-19107](https://issues.apache.org/jira/browse/SPARK-19107), we now can treat hive as a data source and create hive tables with DataFrameWriter and Catalog. However, the support is not completed, there are still some cases we do not support.

This PR implement:
DataFrameWriter.saveAsTable work with hive format with append mode

## How was this patch tested?
unit test added

### CODE REVIEW

**Comment by windpiger** [comment]
retest this please

**Comment by cloud-fan** [comment]
The overall idea is to use `InsertIntable` to implement appending to hive table, but this approach is too hacky, we should follow the way how we deal with data source table, e.g. `DataFrameWriter.saveAsTable` just build a `CreateTable` plan, rule `AnalyzeCreateTable` do some checking and normalization, and another rule turn `CreateTable` into `CreateDataSourceTableAsSelectCommand`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 1]
```diff
@@ -112,20 +112,23 @@ case class AnalyzeCreateTable(sparkSession: SparkSession) extends Rule[LogicalPl
...
+          val specified = DataSource.lookupDataSource(tableDesc.provider.get)
+          (existing == specified, existing.getSimpleName, specified.getSimpleName)
+        case true =>
+          val existing = existingTable.provider.get
+          val specified = tableDesc.provider.get
+          (existing == specified, existing, specified)
       }
-
       // Check if the specified data source match the data source of the existing table.
-      val existingProvider = DataSource.lookupDataSource(existingTable.provider.get)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: We have `HiveFileFormat`, and we can make it implement `DataSourceRegister`, then `DataSource.lookupDataSource("hive")` can work.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala [hunk 2]
```diff
@@ -69,7 +68,7 @@ case class CreateHiveTableAsSelectCommand(
         withFormat
       }
 
-      sparkSession.sessionState.catalog.createTable(withSchema, ignoreIfExists = false)
+      sparkSession.sessionState.catalog.createTable(withSchema, ignoreIfExists = true)
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala]: looks like we don't need to build `withSchema` anymore, the schema will be set in `AnalyzeCreateTable`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 3]
```diff
@@ -1353,6 +1353,15 @@ class HiveDDLSuite
       sql("INSERT INTO t SELECT 2, 'b'")
       checkAnswer(spark.table("t"), Row(9, "x") :: Row(2, "b") :: Nil)
 
+      Seq(10 -> "y").toDF("i", "j")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: please add a new test, to append to a hive table, also test append to a data source table with hive provider and check the error message

**Comment by windpiger** [comment]
retest this please

File: sql/hive/src/main/resources/META-INF/services/org.apache.spark.sql.sources.DataSourceRegister [hunk 4]
```diff
@@ -1 +1,2 @@
 org.apache.spark.sql.hive.orc.OrcFileFormat
+org.apache.spark.sql.hive.execution.HiveFileFormat
```
Comment (cloud-fan) [inline sql/hive/src/main/resources/META-INF/services/org.apache.spark.sql.sources.DataSourceRegister]: is this auto-generated?

**Comment by windpiger** [comment]
add it manual

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala [hunk 5]
```diff
@@ -43,7 +44,18 @@ import org.apache.spark.util.SerializableJobConf
  *
  * TODO: implement the read logic.
  */
-class HiveFileFormat(fileSinkConf: FileSinkDesc) extends FileFormat with Logging {
+class HiveFileFormat
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala]: we should keep the main constructor here, and add a new parameter-less constructor.

**Comment by windpiger** [comment]
if change these two constructor，the paramless
 Constructor will be 
def this() = this(null) or this(new FileSinkDesc(xx))
This is strange, or there is a better way

**Comment by cloud-fan** [comment]
LGTM, pending test

**Comment by windpiger** [comment]
retest this please

**Comment by gatorsmile** [comment]
You need to fetch the upstream and merge it to your local branch. Some changes were made and merged to the upstream/master, although they did not introduce the conflicts. The changes caused the compilation errors in your PR.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala [hunk 6]
```diff
@@ -1167,16 +1161,15 @@ class MetastoreDataSourcesSuite extends QueryTest with SQLTestUtils with TestHiv
...
-      assert(e.contains("Failed to find data source: hive"))
+    withTable(tableName) {
+      sql(
+        s"""
+           |CREATE TEMPORARY VIEW $tableName
+           |(col1 int)
+           |USING hive
+         """.stripMargin)
+      val tempView = spark.sessionState.catalog.getTempView(tableName)
+      assert(tempView.isDefined, "create a temp view using hive should success")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala]: hmmm, it's not expected. let's add a check in `CreateTempViewUsing`, and throw exception for hive provider, e.g. `if (DDLUtils.isHiveTable(t)) throw ...`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 7]
```diff
@@ -112,12 +112,6 @@ case class AnalyzeCreateTable(sparkSession: SparkSession) extends Rule[LogicalPl
         throw new AnalysisException("Saving data into a view is not allowed.")
       }
 
-      if (DDLUtils.isHiveTable(existingTable)) {
-        throw new AnalysisException(s"Saving data in the Hive serde table $tableName is " +
-          "not supported yet. Please use the insertInto() API as an alternative.")
-      }
-
-      // Check if the specified data source match the data source of the existing table.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: why remove this line?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ddl.scala [hunk 8]
```diff
@@ -65,6 +65,10 @@ case class CreateTempViewUsing(
   }
 
   def run(sparkSession: SparkSession): Seq[Row] = {
+    if (provider.toLowerCase == DDLUtils.HIVE_PROVIDER) {
+      throw new AnalysisException("Currently Hive data source can not be created as a view")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ddl.scala]: `Hive data source can only be used with tables, you cannot use it with CREATE TEMP VIEW USING`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 9]
```diff
@@ -1461,6 +1461,25 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
     })
   }
 
+  test("run sql directly on files - hive") {
+    withTable("t") {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: you don't need to create a table
```
withTempPath { path =>
  spark.range(100).toDF.write.parquet(path.getAbsolutePath)
  ...
  sql(s"select id from hive.`${path.getAbsolutePath}`")
}
```

**Comment by windpiger** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ddl.scala [hunk 10]
```diff
@@ -65,6 +65,11 @@ case class CreateTempViewUsing(
   }
 
   def run(sparkSession: SparkSession): Seq[Row] = {
+    if (provider.toLowerCase == DDLUtils.HIVE_PROVIDER) {
+      throw new AnalysisException("Hive data source can not be used with tables," +
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ddl.scala]: `can only be used`

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ddl.scala]: and please add a space after `,`

**Comment by windpiger** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**